activesupport/lib/active_support/hash_with_indifferent_access.rb

```diff
@@ -94,7 +94,11 @@ def self.[](*args)
     #
     # This value can be later fetched using either +:key+ or +'key'+.
     def []=(key, value)
-      regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      if value.frozen?
+        regular_writer(convert_key(key), convert_value(value))
+      else
+        regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      end
```

Comment:
You could have it inside `check_value` if it was combined with the check for `options[:for] == :assignment` or am I missing something?, e.g:

``` ruby
elsif value.is_a?(Array)
  if options[:for] != :assignment || value.frozen?
    value = value.dup
  end
  value.map! { |e| convert_value(e, options) }
else
```

Given the short-circuiting of conditionals it shouldn't have any noticeable effect.